feat(certificate-auth): Added support for PFX certificate authentication

rvelaVenafi opened the pull request:
Closes VC-27220

rvelaVenafi committed:
feat(certificate-auth): Added support for PFX certificate authentication

Closes VC-27220

luispresuelVenafi commented on venafi/provider.go:
```diff
@@ -20,67 +26,116 @@ const (
+	envVenafiClientID       = "VENAFI_CLIENT_ID"
+
+	// Attributes of the provider
+	fURL         = "url"
```
Could you follow this pattern?

`provider + Attr + <attribute name>`

for this case:

`providerAttrUrl`

Edit: I mean `providerAttrUrl` instead of `fURL`

Ref: https://github.com/Venafi/terraform-provider-venafi/blob/master/venafi/resource_venafi_certificate.go#L69

Edit2: I looking at my comment, I see that it was not 100% clear, I was referring to the constant name, not the value in it.

rvelaVenafi replied:
This is already set and released. I just moved it to a constant. I cannot change it without breaking backwards compatibility

luispresuelVenafi replied:
It won't break anything on backwards, it is just renaming the constant, not the values themselves.

Also, how it is that they are set and released if this is the first time you are adding this new constants? If you believe that I'm referring to change the values of the variables, I confirm that I'm not meaning that, rather to change the constant name itself, not the value, that should not promote any backward compatibility issue.

Edit: I looking at my first comment, I see that it was not 100% clear, I was referring to the constant name, not the value in it.

rvelaVenafi replied:
Ok , my bad. You mean the variable names.

Why do I need to create such long names for the variable? Is there any other provider living on this plugin?
I can agree on changin `fUrl` to `attrUrl`, as attribute is the correct name of these variables in terraform context. But I cannot see how adding the provider to all of them makes the code any better. It is just more unnecessary noise as this plugin is specifically for this provider, no need to differentiate. Context gives you all you need to understand what `fURL` or `attrURL` means

luispresuelVenafi replied:
I was adding the provider prefix because we have also other `<resourceName>` + `Attr` + `<attribute Name>`, for now only one, but is my intention we follow this standard in the future in order resources. That said, any of both `providerAttrUrl` or `attrUrl` of both looks good to me. If you like to follow the latter I'm ok with the change.

rvelaVenafi replied:
renamed constants to `provider<Name>`

rvelaVenafi committed:
feat(certificate-auth): Fixed issue with dev mode not working

Closes VC-27220

rvelaVenafi committed:
feat(certificate-auth): Renamed constants

Closes VC-27220

rvelaVenafi committed:
feat(certificate-auth): Updated pkcs12 reading method

Changed attribute p12_cert to p12_cert_filename

Changed attribute p12_password to p12_cert_password

Updated documentation

Closes VC-27220

rvelaVenafi committed:
feat(certificate-auth): Refactored pkcs12 functions

Closes VC-27220

sbanic-venafi commented on venafi/provider.go:
```diff
@@ -94,106 +156,123 @@ Example:
+		Zone:       zone,
+		LogVerbose: true,
+		Credentials: &endpoint.Authentication{
+			ClientId: clientID,
```
This client ID is set for each of the if/else choices below, whereas in previous logic it was not set. Is this acceptable?

luispresuelVenafi replied:
I believe it was missed to mention in the description. Since Russel is integrating the PFX auth, he's also adding the Client ID which I'd have expected to be hardcoded in the terraform code at very least in the frist place like our other Open Source integrations (I just verified that master code doesn't have it). But it seems it was not, so he's also integrating it with default value "`hashicorp-terraform-by-venafi`" but also allowing to edit with the attribute "providerClientID". Could you confirm @rvelaVenafi ?

rvelaVenafi replied:
Indeed, somehow, the provider lacked two things:
1. A default clientID, which means all tokens requested by the provider were issued under `vcert-sdk`  (default of vcert golang SDK) clientID 
2. A way to pass your clientID, which means there is no way to tell TPP who is requesting tokens, as all tokens requested by the provider go with `vcert-sdk` as client

I added ability to pass your own ID as well as set a default ID for the provider

sbanic-venafi replied:
OK, I think I can see it now. Thanks.
